dive_core/perf_metrics_data.cpp

```diff
@@ -0,0 +1,192 @@
+*/
+
+#include "dive_core/perf_metrics_data.h"
+
+#include <cctype>
+#include <cerrno>
+#include <cstdlib>
+#include <iostream>
+#include <filesystem>
+#include <string>
+#include <vector>
+#include <array>
+
+#include "dive_core/available_metrics.h"
+#include "dive_core/common/string_utils.h"
+
+namespace Dive
+{
+namespace
+{
+constexpr std::array kFixedHeaders = { "ContextID",   "ProcessID", "FrameID",
+                                       "CmdBufferID", "DrawID",    "DrawType",
+                                       "DrawLabel",   "ProgramID", "LRZState" };
+
+bool ParseHeaders(const std::string&              line,
+                  const AvailableMetrics&         available_metrics,
+                  std::vector<std::string>&       metric_names,
+                  std::vector<const MetricInfo*>& metric_infos)
+{
+    std::stringstream header_ss(line);
+    std::string       header_field;
+    int               column_index = 0;
+    while (StringUtils::GetTrimmedField(header_ss, header_field, ','))
+    {
+        if (column_index < kFixedHeaders.size())
+        {
+            if (header_field != kFixedHeaders[column_index])
+                return false;
+        }
+        else
+        {
+            metric_names.push_back(header_field);
+            metric_infos.push_back(available_metrics.GetMetricInfo(header_field));
+        }
+        column_index++;
+    }
+
+    return column_index >= kFixedHeaders.size();
+}
+
+std::optional<PerfMetricsRecord> ParseRecordFixedFields(const std::vector<std::string>& fields)
+{
+    if (fields.size() < kFixedHeaders.size())
+    {
+        return std::nullopt;
+    }
+
+    PerfMetricsRecord record{};
+    if (!StringUtils::SafeConvertFromString(fields[0], record.m_context_id) ||
+        !StringUtils::SafeConvertFromString(fields[1], record.m_process_id) ||
+        !StringUtils::SafeConvertFromString(fields[2], record.m_frame_id) ||
+        !StringUtils::SafeConvertFromString(fields[3], record.m_cmd_buffer_id) ||
+        !StringUtils::SafeConvertFromString(fields[4], record.m_draw_id) ||
+        !StringUtils::SafeConvertFromString(fields[5], record.m_draw_type) ||
+        !StringUtils::SafeConvertFromString(fields[6], record.m_draw_label) ||
+        !StringUtils::SafeConvertFromString(fields[7], record.m_program_id) ||
+        !StringUtils::SafeConvertFromString(fields[8], record.m_lrz_state))
+    {
+        return std::nullopt;  // Parsing failed
+    }
+    return record;
+}
+
+template<typename T>
+bool ParseAndEmplaceMetric(const std::string& s, std::vector<std::variant<int64_t, float>>& values)
+{
+    T val{};
+    if (StringUtils::SafeConvertFromString(s, val))
+    {
+        values.emplace_back(val);
+        return true;
+    }
+    return false;
+}
+
+bool ParseMetrics(const std::vector<std::string>&       fields,
+                  const std::vector<const MetricInfo*>& metric_infos,
+                  std::optional<PerfMetricsRecord>&     record_opt)
```

Comment:
Since the record is required, consider taking it by reference. This better communicates the intent to the caller and means that you can't accidentally derefence an empty optional:
```suggestion
                  PerfMetricsRecord&     record)
```
It's easy to pass a reference by dereferencing the optional at the callsite (after has_value() has been checked)
```c++
if (ParseMetrics(fields, metric_infos, *record_opt))
```